### ISSUE TITLE
Task Autoloading

### ISSUE THREAD

**Issue by warslett** [deployphp/deployer#1372]
| Q                 | A
| ----------------- | ---
| Issue Type        | Feature Request
| Deployer Version  | master

### Description 
Our deployment is quite complicated and we have split task definition up amongst multiple files which we store in a repository and pull in with composer. Each time we want to use a task that was defined in one of these files we have to "require" the php file where it is defined. It feels like we could "autoload" these relatively simply. The behaviour would be something like this:

1. A task is called such as `web:bower:install`
2. Deployer would check whether the task has already been defined (maybe defined in the same file)
3. Deployer would attempt to include a file `deployer/web/bower/install.php` where the task _could_ be defined and would check again whether the task is now defined
4. Deployer would attempt to load a file `deployer/web/bower.php` where all the tasks in the deploy:web:bower namespace _could_ be defined and would check again whether the task is now defined.
5. Deployer would keep attempting to load files for each level of the namespace until the task is defined and it can run it. If it can't find the file only then does it fail.

This way you could choose to define a task in a single file deploy/web/bower/install.php or you could choose to define all the tasks in the same namespace in a single file. You could even define an autoload hierarchy so that you can load tasks from both local sources and from the vendor directory. Maybe with a definition in your deploy.php. Something like this:

```
<?php

namespace Deployer;

require_once('vendor/autoload.php');

autoload([ //default configuration
  'deployer/',
  'vendor/'
]);

task('deploy', [
    'vendor-name:package:task', //any task can be overridden by adding a definition to the deployer directory
]);
```

Then pull in the package with `composer require vendor-name/package`. Wanted to get a feel for the communities thoughts on the feature. Would be happy to do the work on it if nobody can see any issues with it.

**Comment by antonmedv** [comment]
I like this feature. But I would like to think about problem what we solving. And what is the best solution for it.

**Comment by warslett** [comment]
@antonmedv thanks for the reply. The problem is essentially the same problem of autoloading with PHP. The simplest benefit is that we no longer have to require a file to use a task defined in it. The effort wasted here is not just the effort of writing a line of code but in actually digging around your source files to figure out where the task you need is defined.

This is a particular issue when it comes to using tasks that have been pulled in from 3rd party composer packages where you won't know the codebase because somebody else wrote it. This is not what composer users are used to: when you pull in a 3rd party library with composer you can use all of the Classes and functions in that package out the box because they are autoloaded. So why not docker tasks? This improvement may actually lead to greater sharing of task definitions across the community: more collaboration and less reinventing the wheel.

Additionally, autoloading is more reliable than manually requiring source files. Let's say I require a file which then requires another file which defines a task I am using in my deploy.php, then one day the require for the file with my task in it get's removed (again, could be an issue for 3rd party packages in particular).

If the tasks are autoloaded, I never have to deal with the problem of relying that a file I need has been required somewhere else, when I need a task, the file get's required. It also means that files with task definitions that are not called on a particular run are not loaded unnecessarily.

The bottom line is, more automation is always better. Nobody wants to do anything manually if they can avoid it.

**Comment by antonmedv** [comment]
Okay, here is what I'm thinking right now:

- Will be cool to autoload task after composer require step automatically.
- Maybe use special field in composer.json file to point to the place there recipes located.

But there are two unsolved problems:

1. How to define order of task loading if two recipes have same task names? Second one will be automatically overriding first one.
2. How to load only needed recipes? For example, _deployer/recipes_ comes with a lot of recipes inside and it will be messy if they all will be loaded.

**Comment by antonmedv** [comment]
Also will be cool to autoload init templates #1340

**Comment by warslett** [comment]
I like the idea of defining the autoload configuration in the composer.json. Something like this would allow a lot of flexibility:
```
{
   ...
    "extra": {
        "deployer-autoload": {
            "deployer/": ":", //root namespace
            "vendor/": ":" //root namespace
        }
    }
}
```
The mappings are ordered in priority from top to bottom so a task definition in "deployer/" will override a task definition in "vendor/". If an autoload configuration is not defined, the default would be to look in deployer/ and vendor/

Autoloading won't work where you have duplicate task names, autoloading (whether it's classes and functions or tasks) relies on identifiers being unique and this problem is usually addressed by namespacing. This means it is impossible to autoload those recipes with their current naming because they use the same task names. I think that for the moment it could be best to treat autoloading the recipes as a separate problem and leave them as they are but in the long run there is something to be said for using proper namespacing with the recipes.

If instead of "deploy:assets:install" this task was named "deployer:symfony:assets:install" then this task can now be autoloaded. You could even split the recipe off into it's own repo. That way I can require only the recipe I need `composer require deployer/symfony`. This would open the door for other users to contribute their own recipes to the community as packages. Appreciate this would be a radical breaking change to the recipes hence why I'd suggest it is considered separately.

Sorry for the long comments, want to make sure I'm being clear about how I envisage it working.

**Comment by antonmedv** [comment]
Again this requires a lot of manual configuring and leaves a few problems and unsolved issues.
I like current process of installing recipes as it has nice properties:
* simple — only two steps: `composer require` and `require 'file.php'`
* native — use only php stuff, no DSL
* every body know what it's working

But if we can figure out something better for developer experience, it will be great. (But I don't want to invent a wheel, maybe take a look on other tools an solutions for this problem). 
For example, only  `composer require` and init script already has new template.

**Comment by warslett** [comment]
@antonmedv I am proposing that we have a default configuration of loading from "deployer/" which is for local task definitions and "vendor/" which is for 3rd party definitions. This will work out the box if we can start to establish a covention for namespacing tasks. This is essentially how all autoloading works. In PHP the convention is PSR-4. Then nobody needs to configure anything because everyone namespaces their tasks the same way. I appreciate that this would be a **radical** change so maybe we park that for a moment. Maybe one simple solution that will help us for the time being is to just introduce a function for registering an auto load closure in the same way as PHP has `spl_autoload_register($callable)` (http://php.net/manual/en/function.spl-autoload-register.php) which is what composer uses. For example:

```
<?php
# deploy.php
namespace Deployer;

registerTaskAutoloader(function($taskName){
   //logic for autoloading task definition goes here
});

localServer('dev')
    ->set('deploy_path', __DIR__)
;

task('deploy', [
    'docker:compose:up' //task not defined in scope but before failing, deployer calls our autoloader and then tries again
]);
```
This would be a relatively minor addition which would allow any developer the option of defining their own behaviour for autoloading or they can carry on calling require if they prefer. This is how autoloading was introduced in PHP before conventions like PSR-4 were established. You won't get the benefit of making it easier to share and collaborate on task definitions until some sort of convention can be established but this initial solution will at least allow us to autoload our tasks in projects where a deployment is 20+ tasks pulled in from multiple private repositories.

**Comment by antonmedv** [comment]
Also what about listing available tasks? Should we list them all?
What about using php include path? Each 3rd party recipe can specify special include path of their recipes.

**Comment by warslett** [comment]
Listing all available tasks could be complicated because a task could be defined in any php file since it can have any name so to find all possible tasks that can be autoloaded, deployer would need to read every single php file in the configured paths and see if they have any tasks with valid names defined. This could be very complicated and a very heavy process.

Setting the include path is still not really autoloading and would still leave the problem of having to rely that the correct file has been required somewhere in the process and having to know what that file is.

The way I see it there are two possible features that could be beneficial to begin with:
1. Provide a registerTaskAutoloader function to allow developers to register their own autoload function that way we aren't imposing any naming/ namespacing conventions on the community just yet and we are delegating the autoloading behaviour to the individual projects. If people don't want to use it they don't have to (just like PHP).
2. The full solution: we build autoloading into Deployer and establish a naming and file structure convention for tasks so that Deployer can figure out what file it needs to load by the task name which would open the door to seamless autoloading from 3rd party packages but establishing conventions could be a radical change. The default configuration would work out the box but there would also be an option to configure autoloading via the composer.json file.

Number 1 would be a good simple starting point that would open up the possibility of autoloading and some sort of autoloading conventions might evolve from the community once this has been implemented which would make way for number 2. I would be happy to create a PR for number 1 to prove the concept.

**Comment by Blizzke** [comment]
Reminds me a bit of what laravel recently did to auto-discover packages (https://laravel-news.com/package-auto-discovery). They turned the wheel around and make the package you are installing responsible for being auto-discoverable. They also make it possible to add configuration to your own `composer.json` to skip discovering a package. The discovery will result in the definition of a serviceprovider that needs to be called. This allows the package to register itslef (this can basically including everything then, not only tasks).

For deployer this could work both ways: Allow external plugins to make themselves discoverable by providing the configuration in their composer.json. For your own tasks you can add the same to the local composer.json. Probably the code they wrote for this can be "borrowed" for a big part.

**Comment by kszymukowicz** [comment]
When someone would like to implement it then please remember that its not only about task file begin included automatically. I can imagine that set of tasks are using common config variables, common options etc. This also needs to be loaded somehow.

**Comment by antonmedv** [comment]
I'm thinking about not-autoloader now, but `import` api.

Something like this:
```php
import('recipe/deploy/vendor.php');
import('github.com/deployphp/deployer/blob/master/recipe/laravel.php');
import('registry.mycorm.com/common.php');
```

If package not found locally, it will be downloaded and stored at `.dep` dir in project root.

**Comment by warslett** [comment]
@antonmedv sounds like you are trying to implement your own dependency management system for deployer. This might be useful but could be very complicated with lots of challenges. Have you considered the usual complexities with dependency management:

- What happens if an import has it's own import? Will deployer import recursively?
- Importing a remote script could introduce unexpected and inconsistent behaviour if that script is updated. This is why most dependency management systems allow you to pin to a specific version so they can manage the introduction of bugs and deprecations from third party packages
- What happens if two imports need to import two different versions of the same package?

These problems are all already solved by Composer. Why reinvent the wheel?

**Comment by antonmedv** [comment]
> What happens if an import has it's own import? Will deployer import recursively?
Yes.

>Importing a remote script could introduce unexpected and inconsistent behaviour if that script is updated. This is why most dependency management systems allow you to pin to a specific version so they can manage the introduction of bugs and deprecations from third party packages
All downloaded recipes will be saved at `.dep` folder and it should be added to git. Also specific versions can be easily used via:
 ```php
import('github.com/deployphp/deployer/blob/0e99b659ba931813ed87f2f8e90caec93f5543ac/recipe/laravel.php');
```
> What happens if two imports need to import two different versions of the same package?
This is a problem right now.

Composer now solving it for deployer for these reasons:
* You can specify package, but no autoloading to recipes.
* Can't use different version of same recipe (same as above).

Actually all this was a pain with `deployer/recipes` package and I'm moving all recipes from there to `deployer/deployer` under `contrib` subdir in v7.

**Comment by antonmedv** [comment]
The new version of Deployer currently developing in master, so it's the best time to introduce new features.

Please, create your PR.
